sql/src/test/java/org/apache/druid/sql/calcite/BaseCalciteQueryTest.java

```diff
@@ -830,163 +760,93 @@ public void testQuery(
-        queryContext,
-        parameters,
-        sql,
-        authenticationResult,
-        createOperatorTable(),
-        createMacroTable(),
-        CalciteTests.TEST_AUTHORIZER_MAPPER,
-        queryJsonMapper
-    );
+    return new QueryTestBuilder(new CalciteTestConfig())
+        .cannotVectorize(cannotVectorize)
+        .skipVectorize(skipVectorize);
   }
 
-  private Pair<RowSignature, List<Object[]>> getResults(
-      final PlannerConfig plannerConfig,
-      final Map<String, Object> queryContext,
-      final List<SqlParameter> parameters,
-      final String sql,
-      final AuthenticationResult authenticationResult,
-      final DruidOperatorTable operatorTable,
-      final ExprMacroTable macroTable,
-      final AuthorizerMapper authorizerMapper,
-      final ObjectMapper objectMapper
-  )
+  public class CalciteTestConfig implements QueryTestBuilder.QueryTestConfig
   {
-    final SqlStatementFactory sqlStatementFactory = getSqlStatementFactory(
-        plannerConfig,
-        new AuthConfig(),
-        operatorTable,
-        macroTable,
-        authorizerMapper,
-        objectMapper
-    );
-    final DirectStatement stmt = sqlStatementFactory.directStatement(
-        SqlQueryPlus.builder(sql)
-            .context(queryContext)
-            .sqlParameters(parameters)
-            .auth(authenticationResult)
-            .build()
-    );
-    Sequence<Object[]> results = stmt.execute().getResults();
-    RelDataType rowType = stmt.prepareResult().getReturnedRowType();
-    return new Pair<>(
-        RowSignatures.fromRelDataType(rowType.getFieldNames(), rowType),
-        results.toList()
-    );
-  }
+    @Override
+    public QueryTestRunner analyze(QueryTestBuilder builder)
+    {
+      if (builder.expectedResultsVerifier == null && builder.expectedResults != null) {
+        builder.expectedResultsVerifier = defaultResultsVerifier(
+            builder.expectedResults,
+            builder.expectedResultSignature
+        );
+      }
+      final List<QueryTestRunner.QueryRunStep> runSteps = new ArrayList<>();
+      final List<QueryTestRunner.QueryVerifyStep> verifySteps = new ArrayList<>();
+      if (builder.expectedResources != null) {
+        Preconditions.checkArgument(
+            builder.expectedResultsVerifier == null,
+            "Cannot check both results and resources"
+        );
+        QueryTestRunner.PrepareQuery execStep = new QueryTestRunner.PrepareQuery(builder);
+        runSteps.add(execStep);
+        verifySteps.add(new QueryTestRunner.VerifyResources(execStep));
+      } else {
+        QueryTestRunner.ExecuteQuery execStep = new QueryTestRunner.ExecuteQuery(builder);
+        runSteps.add(execStep);
+        if (builder.expectedResultsVerifier != null) {
+          verifySteps.add(new QueryTestRunner.VerifyResults(execStep));
+        }
+        if (builder.expectedQueries != null) {
+          verifySteps.add(new QueryTestRunner.VerifyNativeQueries(execStep));
+        }
```

Comment:
while we are here, can we swap this to verify the native query before verifying the results? I know this is how it used to be, but it seems better to verify that the queries are correct before the results are correct :shrug: